pkg/resource/inference_component/hooks.go

```diff
@@ -0,0 +1,166 @@
+// permissions and limitations under the License.
+
+package inference_component
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	ackcompare "github.com/aws-controllers-k8s/runtime/pkg/compare"
+	ackrequeue "github.com/aws-controllers-k8s/runtime/pkg/requeue"
+	svcapitypes "github.com/aws-controllers-k8s/sagemaker-controller/apis/v1alpha1"
+	svccommon "github.com/aws-controllers-k8s/sagemaker-controller/pkg/common"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	svcsdk "github.com/aws/aws-sdk-go/service/sagemaker"
+	"reflect"
+)
+
+var (
+	modifyingStatuses = []string{
+		svcsdk.InferenceComponentStatusCreating,
+		svcsdk.InferenceComponentStatusUpdating,
+		svcsdk.InferenceComponentStatusDeleting,
+	}
+
+	resourceName = GroupKind.Kind
+
+	lastSpecForUpdateAnnotation = fmt.Sprintf("%s/last-spec-for-update", GroupKind.Group)
+
+	requeueWaitWhileDeleting = ackrequeue.NeededAfter(
+		errors.New(resourceName+" is Deleting."),
+		ackrequeue.DefaultRequeueAfterDuration,
+	)
+)
+
+// customDescribeInferenceComponentSetOutput sets the resource ResourceSynced condition to False if
+// InferenceComponent is being modified by AWS
+func (rm *resourceManager) customDescribeInferenceComponentSetOutput(ko *svcapitypes.InferenceComponent) {
+	latestStatus := ko.Status.InferenceComponentStatus
+	svccommon.SetSyncedCondition(&resource{ko}, latestStatus, &resourceName, &modifyingStatuses)
+}
+
+// customUpdateInferenceComponentSetOutput sets ConditionTypeResourceSynced condition to True or False
+// based on the InferenceComponentStatus on AWS so the reconciler can determine if a
+// requeue is needed
+func (rm *resourceManager) customUpdateInferenceComponentSetOutput(ko *svcapitypes.InferenceComponent) error {
+	//set last inference component spec used for update in annotations
+	annotations := ko.ObjectMeta.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+	spec := ko.Spec.DeepCopy()
+	spec.Tags = nil
+	serializedSpec, err := json.Marshal(spec)
+	if err != nil {
+		return err
+	}
+	annotations[lastSpecForUpdateAnnotation] = string(serializedSpec)
+	ko.ObjectMeta.SetAnnotations(annotations)
+
+	// injecting Updating status to keep the Sync condition message and status.InferenceComponentStatus in sync
+	ko.Status.InferenceComponentStatus = aws.String(svcsdk.InferenceComponentStatusUpdating)
+
+	latestStatus := ko.Status.InferenceComponentStatus
+	svccommon.SetSyncedCondition(&resource{ko}, latestStatus, &resourceName, &modifyingStatuses)
+
+	return nil
+}
+
+// customUpdateInferenceComponentPreChecks adds specialized logic to check if controller should
+// proceed with UpdateInferenceComponent call.
+// Update is blocked in the following cases:
+//  1. while InferenceComponentStatus != InService (handled by requeueUntilCanModify method).
+//  2. InferenceComponentStatus == Failed.
+//  3. A previous update to the InferenceComponent with same spec failed.
```

Comment:
Yes, that is the reason for checking this condition. Since InferenceComponent is a Blue/Green deployment if the Update fails it will rollback to last successful state with Status = InService. So ACK will re-trigger the update. After discussion with product it was decided to move the resource to terminal state if an update fails.